### PR TITLE
Expand KlayCredentials for multiSig and rolebasedAccount

### PR DESCRIPTION
**Pull request by nogang**
## Proposed changes

- Caver-java needs to extend the KlayCredentials you use for multisig, roleBased account
- To sign a rolebased account, KlayCredentials must have a separate set of corresponding transactions, updates, and feePayers.
- KlayCredentials must be able to hold multiple privateKeys for multisig.
## Types of changes

Please put an x in the boxes related to your change.

- [ ] Bugfix
- [x] New feature or enhancement
- [ ] Others

## Checklist

*Put an x in the boxes that apply. You can also fill these out after creating the PR. If you're unsure about any of them, don't hesitate to ask. We're here to help! This is simply a reminder of what we are going to look for before merging your code.*

- [x] I have read the [CONTRIBUTING GUIDELINES](https://github.com/klaytn/caver-java/blob/master/CONTRIBUTING.md) doc
- [x] I have signed the [CLA](https://cla-assistant.io/klaytn/caver-java)
- [x] Lint and unit tests pass locally with my changes
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] I have added necessary documentation (if appropriate)
- [ ] Any dependent changes have been merged and published in downstream modules

## Related issues

- Please leave the issue numbers or links related to this PR here.

## Further comments

If this is a relatively large or complex change, kick off the discussion by explaining why you chose the solution you did and what alternatives you considered, etc...

### CODE REVIEW

File: core/src/main/java/com/klaytn/caver/crypto/KlayCredentials.java [hunk 1]
```diff
@@ -24,20 +24,61 @@
...
+        this.ecKeyPairForUpdateList = (ecKeyPairForUpdateArray != null && ecKeyPairForUpdateArray.size() != 0) ? Collections.unmodifiableList(ecKeyPairForUpdateArray) : null;
+        this.ecKeyPairForFeeFeePayerList = (ecKeyPairForFeeArray != null && ecKeyPairForFeeArray.size() != 0) ? Collections.unmodifiableList(ecKeyPairForFeeArray) : null;
         this.address = address;
     }
 
     public ECKeyPair getEcKeyPair() {
-        return ecKeyPair;
+        if (ecKeyPairForTransactionList != null && ecKeyPairForTransactionList.size() > 0) {
+            return ecKeyPairForTransactionList.get(0);
+        } else {
```
Comment (kjhman21) [inline core/src/main/java/com/klaytn/caver/crypto/KlayCredentials.java]: I think we don't need `else` here like the below function.

File: core/src/main/java/com/klaytn/caver/crypto/KlayCredentials.java [hunk 2]
```diff
@@ -24,20 +24,61 @@
...
 
     private KlayCredentials(ECKeyPair ecKeyPair, String address) {
-        this.ecKeyPair = ecKeyPair;
+        this.ecKeyPairForTransactionList = Collections.unmodifiableList(Arrays.asList(ecKeyPair));
+        this.ecKeyPairForUpdateList = null;
+        this.ecKeyPairForFeeFeePayerList = null;
+        this.address = address;
+    }
+
+    private KlayCredentials(List<ECKeyPair> ecKeyPairForTransactionArray, List<ECKeyPair> ecKeyPairForUpdateArray, List<ECKeyPair> ecKeyPairForFeeArray, String address) {
```
Comment (kjhman21) [inline core/src/main/java/com/klaytn/caver/crypto/KlayCredentials.java]:
```suggestion
    private KlayCredentials(List<ECKeyPair> ecKeyPairForTransactionList, List<ECKeyPair> ecKeyPairForUpdateList, List<ECKeyPair> ecKeyPairForFeeList, String address) {
```

File: core/src/main/java/com/klaytn/caver/crypto/KlayCredentials.java [hunk 3]
```diff
@@ -63,6 +104,52 @@ public static KlayCredentials create(ECKeyPair ecKeyPair, String address) {
...
+    public static KlayCredentials create(List<ECKeyPair> ecKeyPair, String address) {
+        return new KlayCredentials(ecKeyPair, null, null, address);
+    }
+
+    public static KlayCredentials create(List<ECKeyPair> ecKeyPairForTransaction, List<ECKeyPair> ecKeyPairForUpdate, List<ECKeyPair> ecKeyPairForFeePayer, String address) {
+        return new KlayCredentials(ecKeyPairForTransaction, ecKeyPairForUpdate, ecKeyPairForFeePayer, address);
+    }
+
+    public static KlayCredentials create(ECKeyPair[] ecKeyPairsArrayForTransaction, ECKeyPair[] ecKeyPairsArrayForUpdate, ECKeyPair[] ecKeyPairsArrayForFee, String address) {
+        List<ECKeyPair> ecKeyPairsForTransaction = new ArrayList<>(Arrays.asList(ecKeyPairsArrayForTransaction));
```
Comment (kjhman21) [inline core/src/main/java/com/klaytn/caver/crypto/KlayCredentials.java]: Is this really necessary? `new ArrayList<>`?

**Comment by kjhman21** [comment]
Please add comments if possible for all externally visible functions.

File: core/src/main/java/com/klaytn/caver/crypto/KlayCredentials.java (hunk 3 above)
Comment (nogang) [reply core/src/main/java/com/klaytn/caver/crypto/KlayCredentials.java]: You're right. Confused the grammar.

File: core/src/main/java/com/klaytn/caver/crypto/KlayCredentials.java [hunk 4]
```diff
@@ -24,20 +24,60 @@
...
+    private KlayCredentials(List<ECKeyPair> ecKeyPairForTransaction, List<ECKeyPair> ecKeyPairForUpdate, List<ECKeyPair> ecKeyPairForFee, String address) {
+        this.ecKeyPairForTransactionList = (ecKeyPairForTransactionArray != null && ecKeyPairForTransactionArray.size() != 0) ? Collections.unmodifiableList(ecKeyPairForTransactionArray) : null;
+        this.ecKeyPairForUpdateList = (ecKeyPairForUpdateArray != null && ecKeyPairForUpdateArray.size() != 0) ? Collections.unmodifiableList(ecKeyPairForUpdateArray) : null;
+        this.ecKeyPairForFeeFeePayerList = (ecKeyPairForFeeArray != null && ecKeyPairForFeeArray.size() != 0) ? Collections.unmodifiableList(ecKeyPairForFeeArray) : null;
         this.address = address;
     }
 
     public ECKeyPair getEcKeyPair() {
-        return ecKeyPair;
+        if (ecKeyPairForTransactionList != null && ecKeyPairForTransactionList.size() > 0) {
```
Comment (leejh903) [inline core/src/main/java/com/klaytn/caver/crypto/KlayCredentials.java]: How about extract this code `[List] != null && [List].size() > 0` to method? There are many duplications!

File: core/src/main/java/com/klaytn/caver/crypto/KlayCredentials.java [hunk 5]
```diff
@@ -24,20 +24,60 @@
...
-    private final ECKeyPair ecKeyPair;
+    private final List<ECKeyPair> ecKeyPairForTransactionList;
+    private final List<ECKeyPair> ecKeyPairForUpdateList;
+    private final List<ECKeyPair> ecKeyPairForFeeFeePayerList;
     private final String address;
 
     private KlayCredentials(ECKeyPair ecKeyPair, String address) {
-        this.ecKeyPair = ecKeyPair;
+        this.ecKeyPairForTransactionList = Collections.unmodifiableList(Arrays.asList(ecKeyPair));
+        this.ecKeyPairForUpdateList = null;
```
Comment (leejh903) [inline core/src/main/java/com/klaytn/caver/crypto/KlayCredentials.java]: Once we set `null` to variable, we always check `null` if this variable is null or not. How about setting empty list using `collections.emptylist()`?

File: core/src/main/java/com/klaytn/caver/crypto/KlayCredentials.java [hunk 6]
```diff
@@ -63,6 +103,52 @@ public static KlayCredentials create(ECKeyPair ecKeyPair, String address) {
         return new KlayCredentials(ecKeyPair, address);
     }
 
+    public static KlayCredentials create(List<ECKeyPair> ecKeyPair, String address) {
```
Comment (leejh903) [inline core/src/main/java/com/klaytn/caver/crypto/KlayCredentials.java]: It seems Builder pattern can be adjust to this class. This pattern is really effective when there are many combination of creating certain object.

**Comment by nogang** [comment]
> Please add comments if possible for all externally visible functions.


I didn't comment because I didn't comment on the existing code. I will add a comment.

File: core/src/main/java/com/klaytn/caver/crypto/KlayCredentials.java [hunk 7]
```diff
@@ -87,17 +173,34 @@ public boolean equals(Object o) {
 
         KlayCredentials that = (KlayCredentials) o;
 
-        if (ecKeyPair != null ? !ecKeyPair.equals(that.ecKeyPair) : that.ecKeyPair != null) {
+        if (!Objects.equals(ecKeyPairForTransactionList, that.ecKeyPairForTransactionList)) {
```
Comment (leejh903) [inline core/src/main/java/com/klaytn/caver/crypto/KlayCredentials.java]: If there are two `KlayCredentials` objects which have same ecKeyPairs and different order, these objects are not same?

**Comment by kjhman21** [comment]
> I didn't comment because I didn't comment on the existing code. I will add a comment.

For this PR, you don't need to add comments to all functions. Please add comments of newly added functions by this PR.

File: core/src/main/java/com/klaytn/caver/crypto/KlayCredentials.java (hunk 5 above)
Comment (nogang) [reply core/src/main/java/com/klaytn/caver/crypto/KlayCredentials.java]: Thanks good point.

File: core/src/main/java/com/klaytn/caver/crypto/KlayCredentials.java (hunk 6 above)
Comment (nogang) [reply core/src/main/java/com/klaytn/caver/crypto/KlayCredentials.java]: Applying build patten is a good idea. However, this implementation will go as closely as possible to existing usage. If rearchitecturing is done, then consider.

File: core/src/main/java/com/klaytn/caver/crypto/KlayCredentials.java (hunk 7 above)
Comment (nogang) [reply core/src/main/java/com/klaytn/caver/crypto/KlayCredentials.java]: Yes. The order of the signatures differs depending on the order of the keys, so different key sequences are determined by different KlayCredentials.

File: core/src/main/java/com/klaytn/caver/crypto/KlayCredentials.java (hunk 4 above)
Comment (nogang) [reply core/src/main/java/com/klaytn/caver/crypto/KlayCredentials.java]: I'm going to delete the conditional statement that compares null. Should I implement it as a method?

File: core/src/main/java/com/klaytn/caver/crypto/KlayCredentials.java [hunk 8]
```diff
@@ -24,45 +24,219 @@
...
+    /**
+     * Static method for creating KlayCredentials instance
+     *
+     * @param privateKeyArrayForTransaction private key array for transaction signing
+     * @param privateKeyArrayForUpdate private key array for transaction signing
+     * @param privateKeyArrayForFee private key array for transaction signing
+     * @param address address of account
+     * @return KlayCredentials
+     */
+    public static KlayCredentials create(String[] privateKeyArrayForTransaction, String[] privateKeyArrayForUpdate, String[] privateKeyArrayForFee, String address) {
```
Comment (kjhman21) [inline core/src/main/java/com/klaytn/caver/crypto/KlayCredentials.java]:
```suggestion
    public static KlayCredentials create(String[] privateKeyArrayForTransaction, String[] privateKeyArrayForUpdate, String[] privateKeyArrayForFeePayer, String address) {
```

File: core/src/main/java/com/klaytn/caver/crypto/KlayCredentials.java [hunk 9]
```diff
@@ -24,45 +24,219 @@
...
+
+        for (String privateKey : privateKeyArrayForTransaction) {
+            ecKeyPairsForTransaction.add(ECKeyPair.create(Numeric.toBigInt(privateKey)));
+        }
+
+        for (String privateKey : privateKeyArrayForUpdate) {
+            ecKeyPairsForUpdate.add(ECKeyPair.create(Numeric.toBigInt(privateKey)));
+        }
+
+        for (String privateKey : privateKeyArrayForFee) {
```
Comment (kjhman21) [inline core/src/main/java/com/klaytn/caver/crypto/KlayCredentials.java]:
```suggestion
        for (String privateKey : privateKeyArrayForFeePayer) {
```

File: core/src/main/java/com/klaytn/caver/crypto/KlayCredentials.java [hunk 10]
```diff
@@ -87,17 +261,34 @@ public boolean equals(Object o) {
...
+        }
+
+        return Objects.equals(address, that.address);
     }
 
     @Override
     public int hashCode() {
-        int result = ecKeyPair != null ? ecKeyPair.hashCode() : 0;
-        result = 31 * result + (address != null ? address.hashCode() : 0);
+        int transactionHashCode = ecKeyPairForTransactionList != null ? ecKeyPairForTransactionList.hashCode() : 0;
```
Comment (kjhman21) [inline core/src/main/java/com/klaytn/caver/crypto/KlayCredentials.java]: Do we need to check null here?

File: core/src/main/java/com/klaytn/caver/crypto/KlayCredentials.java (hunk 4 above)
Comment (leejh903) [reply core/src/main/java/com/klaytn/caver/crypto/KlayCredentials.java]: If you do not use this condition, I think you do not need to extract them.

**Pull request merged**